core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -87,13 +85,16 @@ class TransactionStateManager(brokerId: Int,
   /** number of partitions for the transaction log topic */
   private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
 
-  // this is best-effort expiration and hence not grabing the lock on metadata upon checking its state
+  // this is best-effort expiration and hence not grabbing the lock on the metadata object upon checking its state
```

Comment:
Hmm, transactionsToExpire() should read the most up to date txnMetadata.txnStartTimestamp. So, we probably still need to synchronize on txnMetadata.